### ISSUE TITLE
Global styles does not work in React.StrictMode

### ISSUE THREAD

**Issue by perjansson** [styled-components/styled-components#3008]
## Environment

System:
OS: macOS 10.15.1
CPU: (8) x64 Intel(R) Core(TM) i7-8569U CPU @ 2.80GHz
Memory: 857.00 MB / 16.00 GB
Shell: 3.2.57 - /bin/bash
Binaries:
Node: 10.15.3 - ~/.nvm/versions/node/v10.15.3/bin/node
Yarn: 1.16.0 - /usr/local/bin/yarn
npm: 6.4.1 - ~/.nvm/versions/node/v10.15.3/bin/npm
npmPackages:
styled-components: ^5.0.1 => 5.0.1
Reproduction
https://codesandbox.io/s/wonderful-hill-6glof

## Steps to reproduce

Load the code sandbox
Click the button "to black" and then click again "to white" and notice how the background did not change back to black.

## Expected Behavior

The should change between black and white

## Actual Behavior

The background stays black after the first press on the toggle button. The global CSS (rendered by GlobalStyle in ComponentBlack) is added twice and only one is removed when unmounting when in Strict Mode. This is because to Strict Mode "intentionally double-invoking" some lifecycle methods to determine side effects that could be a problem in the upcoming async mode in React. When concurrent mode becomes available and React might call render function multiple times the side effect of setting global styles on render will cause issues like this.

As far as I can tell `styled-components` should be `React.Strict` compatible and using the `createGlobalStyle` util is the recommended way to do global styling with `styled-components`, so this seems to be a bug to me.

**Comment by probablyup** [comment]
Hmm.

**Comment by kitten** [comment]
@probablyup this may be related to the allocation ID. I suppose we don’t deduplicate if the content is identical which should fix this issue 😅

**Comment by wadamek65** [comment]
Hi, I've investigated this a little more and I think I found the exact cause of this issue.

The global style component uses a ref to hold it's instance and if it's a null it gets assigned a new instance value.

```
    const instanceRef = useRef(null);

    if (instanceRef.current === null) {
      instanceRef.current = styleSheet.allocateGSInstance(styledComponentId);
    }
```

This is all fine because there is a `useEffect` that removes all styles on component dismount.

```
    useEffect(() => () => globalStyle.removeStyles(instance, styleSheet), EMPTY_ARRAY);
```

In `React.StrictMode` this function returned in the `useEffect` doesn't run when strict mode "mounts" the component a second time which causes the issue because the `instance.value` increments, but previous styles do not get removed.

I feel like this issue is specific to `React.StrictMode`? Is there ever a scenario that the returned callback from `useEffect` would not run when the component dismounts (even in concurrent mode)?

I'm not entirely sure what the best approach to fix this would be but I would be happy to submit a PR if someone gives me a little guidance. Perhaps the instance IDs of the global style components should not be held as a ref but in the enclosing `createGlobalStyle` function? Or detect that the strict mode is on and act accordingly if this is strict mode specific issue?

**Comment by joshamos** [comment]
@wadamek65 I wonder if this is causing https://github.com/styled-components/styled-components/issues/3022 in non strict mode. 🤔

**Comment by wadamek65** [comment]
@joshamos unfortunately I am not familiar with Next, but I believe it should be easy to find out whether or not this causes the leak. If you check out the sandbox that @perjansson and toggle the theme multiple times you will see that the `background-color` property is multiplied in the `<style>` created by `styled-components` (see image below).

![image](https://user-images.githubusercontent.com/36732892/75855125-d6c5a500-5df1-11ea-8ddc-2507a30a62de.png)

If this is the cause of memory leak, same thing might be observable in nextjs maybe?

**Comment by ghengeveld** [comment]
The issue in #3022 doesn't just apply to Next.js, it also affects Jest tests and custom SSR solutions.

**Comment by laurentsmohr** [comment]
Experiencing the same issue, when using React Strictmode

**Comment by codeaid** [comment]
Same here. Ended up moving global styles outside of `React.Strict`:

```
<GlobalStyle />
<React.StrictMode>
    <App />
</React.StrictMode>
```
instead of 
```
<React.StrictMode>
    <GlobalStyle />
    <App />
</React.StrictMode>
```

**Comment by umitkucuk** [comment]
Same here with v5.1.1.
Experiencing a global style duplication issue.

![image](https://user-images.githubusercontent.com/24843599/88476308-fc3d8180-cf3f-11ea-9a3a-d01b11211202.png)

**Comment by eps1lon** [comment]
> In React.StrictMode this function returned in the useEffect doesn't run when strict mode "mounts" the component a second time which causes the issue because the instance.value increments, but previous styles do not get removed.

React doesn't mount the component twice. It renders it twice. Only for the committed render an effect will run (and its cleanup function after that commit is discarded).

> I feel like this issue is specific to React.StrictMode? Is there ever a scenario that the returned callback from useEffect would not run when the component dismounts (even in concurrent mode)?

Today you can reproduce this realiably if you have a component in a suspense boundary where a sibling suspends i.e. the component renders, the tree suspends, react discards the render and does not commit it:  https://codesandbox.io/s/concurrent-mode-suspended-component-commit-lnndy

This can also happen if a low priority render is interrupted by a high priority update but that's a bit harder to reproduce reliably.

Be careful with actually doing side-effects in that demo and logging them. `react@experimental` in DEV mode will still invoke renders twice but will disable logging for the second invocation. Logging states affected by side-effects is only reasonable in production builds.

**Comment by probablyup** [comment]
Can y'all try out `styled-components@test` and let me know if it resolves the StrictMode issue?

**Comment by umitkucuk** [comment]
> Can y'all try out `styled-components@test` and let me know if it resolves the StrictMode issue?

`styled-components@5.2.0-test.10` fixes the issue for me, good job. Cheers! 🍻

**Comment by ardaogulcan** [comment]
Hi, I can see it has been a while but I think I am facing a similar issue with the latest styled-components and latest nextjs with `reactStrictMode: true` defined in `next.config.js`


### Here are the reproducing steps;

 Go to latest example provided by nextjs with styled components https://github.com/vercel/next.js/tree/master/examples/with-styled-components
 
You can install project with the following script
 
 ```
 npx create-next-app --example with-styled-components with-styled-components-app
 ```
 
Once porject is created go to top level and add `next.config.js` with the below content

```js
module.exports = { 
  reactStrictMode: true,
}

```

Run development server by

```
npm run dev
```

Check `body` styles and styled components `style` tag in `head`. You will see body styles are duplciated.

![image](https://user-images.githubusercontent.com/4533960/119873431-d3ea2f00-bf24-11eb-98c3-bc4e79d5124f.png)


### Environment

System:
Mac Mini (M1, 2020)
OS: macOS BigSur 11.4
CPU: Apple M1

Shell: zsh 5.8 (x86_64-apple-darwin20.0)
Binaries:
Node: v16.0.0
npm: v7.10.0
npmPackages:

```json
{
  "next": "latest",
  "react": "^17.0.2",
  "react-dom": "^17.0.2",
  "react-is": "^17.0.2",
  "styled-components": "^5.2.3",
  "babel-plugin-styled-components": "^1.12.0"
}
```

**Comment by edarioq** [comment]
This is horrible and makes me wonder if styled components can be used in a real project..

For anyone still having issues I solved it by setting `reactStrictMode: false,`  in next.config.js and manually applying strict mode the following way:

```
<ThemeProvider theme={theme}>
      <Fonts />
      <GlobalStyle />
      <React.StrictMode>
        <Header />
        <main>
          <Component {...pageProps} />
        </main>
        <Footer />
      </React.StrictMode>
    </ThemeProvider>
    ```

**Comment by tr1s** [comment]
Yeah I started a fresh Next.js app with `npx create-next-app@latest` which now sets strict mode to true in the `next.config.js` (it never did this before).

I then installed styled-components how I usually do and was absolutely losing my mind trying to figure out why my global styles weren't applying in the `_app.js`. Turns out turning off strict mode fixes the problem.

Shortly after I then turned strict mode back on and now global styles are working with it on... so basically I'm completely clueless and don't know what's happening. Just thought I'd type this for reference.

**Comment by quantizor** [comment]
For anyone still encountering this class of issue: a comprehensive fix has been published in `styled-components@6.4.0-prerelease.2`. The root cause — per-instance group allocation in `createGlobalStyle` — has been replaced with a shared-group architecture. Please test and report any regressions.

```
npm install styled-components@6.4.0-prerelease.2
```

See PR #5696 for details.
